### ISSUE TITLE
cheritest test_sandbox_inflate_zeros fails with: "expected to consume 33 bytes, got 0"

### ISSUE THREAD

**Issue by rwatson** [CTSRD-CHERI/cheribsd#90]
Running ```cheritest``` on a recent CheriBSD + 20151218 bit file, all tests pass as expected except for ```test_sandbox_inflate_zeros```:
```
> cheritest test_sandbox_inflate_zeros
TEST: test_sandbox_inflate_zeros: Inflate a compressed buffer of zeros
FAIL: test_sandbox_inflate_zeros: expected to consume 33 bytes, got 0
```

Console output reports a problem with a pointer dereference:
```
cpuid = 0
CHERI cause: ExcCode: 0x02 RegNum: $c01 (tag violation)
$c00: v:1 s:0 p:0000003d b:0000000160101000 l:000000000022f000 o:0 t:0
$c01: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:2f000 t:0
$c02: v:1 s:0 p:0000003d b:0000000160101000 l:000000000022f000 o:2ae20 t:0
$c03: v:1 s:0 p:0000007c b:00000001601008bc l:0000000000000004 o:0 t:0
$c04: v:1 s:0 p:0000007c b:00000001601008b0 l:0000000000000008 o:0 t:0
$c05: v:1 s:0 p:0000007c b:00000001601008ac l:0000000000000004 o:0 t:0
$c06: v:1 s:0 p:0000003d b:0000000160101000 l:000000000022f000 o:2a080 t:0
$c07: v:1 s:0 p:0000007c b:0000000160100c18 l:0000000000000004 o:0 t:0
$c08: v:1 s:0 p:0000007c b:0000000160100be0 l:0000000000000020 o:0 t:0
$c09: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c10: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c11: v:1 s:0 p:0000007c b:00000001600e1000 l:0000000000020000 o:0 t:0
$c12: v:1 s:0 p:0000007c b:0000000160100900 l:0000000000000020 o:0 t:0
$c13: v:1 s:0 p:00000004 b:000000016083d400 l:00000000000000f0 o:0 t:0
$c14: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c15: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c16: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c17: v:1 s:0 p:00000007 b:00000001600b6000 l:000000000002ae30 o:1bae0 t:0
$c18: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c19: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c20: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c21: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c22: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c23: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c24: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c26: v:1 s:0 p:0000003d b:0000000160101000 l:000000000022f000 o:0 t:0
$c31: v:1 s:0 p:00000007 b:00000001600b6000 l:000000000002ae30 o:1ddd8 t:0
Apr 18 18:48:42 beri1 kernel: USER_CHERI_EXCEPTION: pid 1349 tid 100065 (cheritest), uid 999: CP2 fault (type 0x32)
Apr 18 18:48:42 beri1 kernel: Trapframe Register Dump:
Apr 18 18:48:42 beri1 kernel: zero: 0   at: 0x200000    v0: 0x32ec0     v1: 0
Apr 18 18:48:42 beri1 kernel: a0: 0x1f7b8       a1: 0x1f8a0     a2: 0xfffffffffffff000  a3: 0x1f8a0
Apr 18 18:48:42 beri1 kernel: a4: 0     a5: 0   a6: 0   a6: 0
Apr 18 18:48:42 beri1 kernel: t0: 0x1000        t1: 0xffffffffffffffff  t2: 0   t3: 0
Apr 18 18:48:42 beri1 kernel: t8: 0     t9: 0x1f8b0     s0: 0   s1: 0
Apr 18 18:48:42 beri1 kernel: s2: 0     s3: 0   s4: 0   s5: 0
Apr 18 18:48:42 beri1 kernel: s6: 0     s7: 0   k0: 0   k1: 0
Apr 18 18:48:42 beri1 kernel: gp: 0x32ec0       sp: 0x1f720     s8: 0x1f720     ra: 0x8268
Apr 18 18:48:42 beri1 kernel: sr: 0x4000fcf3    mullo: 0x1ca0   mulhi: 0        badvaddr: 0x16012b080
Apr 18 18:48:42 beri1 kernel: cause: 0x148      pc: 0x1ddd8
```

$c1 is indeed untagged, and looks a bit like an integer version of an in-sandbox pointer (i.e., without base/bounds/tag/permissions).

**Comment by rwatson** [comment]
It looks like this may be another case where recent compiler changes have caused ```(void *)``` casts on integers or integer variables to no longer be interpreted as $c0-relative pointers -- in this case, ```__sb_heapbase``` in ```libc_cheri```.

**Comment by rwatson** [comment]
Fixed in c5f8eea427273dd1a36dc4e977553a45de69267d: we can no longer cast directly from ```_sb_heapbase``` to a ```(void *)```. In the future, we should just have ```_sb_heapcap```.

**Comment by rwatson** [comment]
Tagged as CheriCCompat on the basis that other programs may use similar tricks to expose what are effectively virtual addresses via variables rather than symbols -- not a great idiom, but not an entirely unreasonable one.
